### ISSUE TITLE
A proper type system for Nu

### ISSUE THREAD

**Issue by htngr** [nushell/nushell#11108]
### Related problem

We at [Codchi](https://codchi.dev) are currently considering a rewrite from Haskell to Nushell. Coming from Haskell, however, I miss a few things which are crucial for a typed functional language, especially if you want to write more than just a script. Don't get me wrong, Nushell is already an amazing shell, but there are some enhancements that would enable writing larger applications in Nushell.

### Algebraic Datatypes

A really useful feature of Rust and Haskell is *Algebraic Datatypes*, a fancy name for records and sum types. The latter is known as `enum` in Rust and is used for `Result`, `Option` and so on. This is one of the Features which set Rust and Haskell apart from more mainstream languages like Java, Go, C# and so on.

Nushell already has `record<...>` and `table<...>`, but lacks a convenient way to define them (see #10630). 
🎯 I think an ergonomic solution is to allow multiline record declarations and standalone types, just like `struct` in Rust:
```nushell
type foo = record<
    bar: int,
    baz: string,
    my_field: my_custom_type,
    my_other_field: my_alias
    ...
>
type my_custom_type = record<...>
type my_alias = decimal
```

On the other hand, Nushell lacks sum types. The syntax could look like this:
```nushell
type my_sumtype = enum<
  variant_a: int,
  variant_b: record<foo: string, bar: int>
>
```
Also they require pattern matching support to force handling *all* possible variants of a sum type:
```nushell
def handle [x: my_sumtype] {
  match $x {
    variant_a value => { print ($value * 2) }
    variant_b r => { print $"(r.foo): (r.bar)" }
  }
}
> handle (variant_a 21)
42
> handle (variant_b {foo: "Foo", bar: 123 })
Foo: 123
```

### Type checking a Nushell program

Currently (as of 0.87.0) Nushell allows incorrect programs and throws errors only at runtime. This is obviously bad for application development. Consider the following examples:
```nushell
> def foo [x: record<bar: int>] { $x.baz } # ❌ no error but obviously incorrect
> foo { baz: 2 } # ✅ correctly throws a type error
Error: nu::parser::type_mismatch
  × Type mismatch.
   ╭─[entry #16:1:1]
 1 │ foo { baz: 2 }
   ·     ─────┬────
   ·          ╰── expected record<bar: int>, found record<baz: int>
   ╰────

> foo { bar: 2 } # 💥 runtime error...
Error: nu::shell::name_not_found
  × Name not found
   ╭─[entry #13:1:1]
 1 │ def foo [x: record<bar: int>] { $x.baz }
   ·                                    ─┬─
   ·                                     ╰── did you mean 'bar'?
   ╰────

```
Also:
```nushell
> let mod10 = {|x: int| $x mod 10 }
> do $mod10 25
5
> let x = do $mod10 "asd" # ❌ no type error...
> $x # 💥 runtime error
Error: nu::shell::type_mismatch
  × Type mismatch during operation.
   ╭─[entry #1:1:1]
 1 │ let mod10 = {|x: int| $x mod 10 }
   ·                       ─┬ ─┬─ ─┬
   ·                        │  │   ╰── int
   ·                        │  ╰── type mismatch for operator
   ·                        ╰── string
   ╰────
```
🎯 I think Nushell should try to push errors from runtime to type checking as much as possible, this is the point of static typing after all. This would also benefit the [integrated language server #10794](https://github.com/nushell/nushell/issues/10794) which could offer better completions and diagnostics and speedup the development as errors are shown directly in the IDE and not at runtime. Type inference would enhance this even more, supporting automated type propagation throughout the program, which would enhance correctness in the whole application overall.

### Parametric polymorphism (Generics in Rust)

If you `describe` `$mod10` from the previous example, Nushell prints just `closure`. There are already generics in Nu (`list<int>`, `record<foo: string>`, ...).
🎯 Why not extend this to `closure<a, b>` for a closure from `a` to `b` (`$mod10` would get the type `closure<int, int>`)? This would preferably be extended to user defined types. One then could write something like:
```nushell
type result<a> = enum<
    ok a,
    err string
>
def unwrap<a> []: result<a> -> a {
    match $in {
        ok x => { x }
        err msg => { error make { msg: $msg } }
    }
}
```
Also this enables typed higher order functions:
```nushell
def map<a,b> [f: closure<a,b>]: list<a> -> list<b> { ... }
def compose<a,b,c> [g: closure<b,c>, f: closure<a,b>]: a -> c { ... }
```
> 🚀 Even better, although probably unrealistic for a shell: Add traits (Rust) / type classes (haskell). 

### TL;DR

Implement following features from typed functional languages like Haskell to make Nushell ready for application development:
1. Algebraic Datatypes (Enums / Sum types and standalone type definitions)
2. Strictly enforce types in a Nushell program (Pushing runtime errors to type errors)
3. Add generics to `closure` and allow them in user defined types (see 1.)

### Conclusion
All these things are well studied in PL research, which Nushell could imensly benefit from. I think (although I'm not an expert in this field) Nushell could be based on gradual[^1] [Hindley–Milner](https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system) typing. This would also enable full program type inference[^2]. 
I know this is a lot and probably brings changes to the internals, but as Nushell is approaching 1.0 (which will be backwards-compatible from then on, AFAIK) these are important features which would future-proof Nushell and drive its adoption, especially among FP / Nix people.

[^1]: [Gradual typing](https://en.wikipedia.org/wiki/Gradual_typing) is a mix of dynamic and static typing. The best known example is Typescript
[^2]: Yusuke Miyazaki, Taro Sekiyama, and Atsushi Igarashi. 2019. Dynamic type inference for gradual Hindley–Milner typing. <https://doi.org/10.1145/3290331>

**Comment by heinwol** [comment]
I agree that more robust and expressive typesystem could bring a lot to nushell. This move will make it comparable with general purpose production languages. At the same time, nushell won't lose its simplicity: small scripts will still be small scripts, 0 overhead.

The main reasons I don't use nushell everywhere include, yeah, lack of proper generics and sum types (the other being lack of async commands).

tldr: it would be **awesome** if this was implemented

**Comment by amtoine** [comment]
we have some topics already in the queue for tomorrow's meeting, but this topic is in there.
we'd like to talk about it if we get time :pray:

**Comment by ysthakur** [comment]
> TypeScript's type system is turing complete and therefore unsound

I just want to point out that a type system being Turing-complete is not a bad thing. Even Java's type system (which is extremely restrictive) is Turing-complete. Getting expressivity at the cost of unsoundness sounds like an acceptable tradeoff. You can always have a recursion limit or timeout for the type-checking phase. Those are things you might want even with a sound type system, because being Turing-incomplete doesn't mean that you can't accidentally encode a program into your types that takes a billion years to run :)

TypeScript's type system honestly seems well-suited to an interpreted language where you can ignore types as much as you want. Combining gradual typing with System F may be an ongoing area of research, but TypeScript's been used in the field for years, and successfully. It might be worth borrowing from.

**Comment by KAAtheWiseGit** [comment]
One thing I'd like to add is that Nushell already has poor man's sum types in form of `OneOf`.  And while a complete type system might be complex, currently most of that complexity is rolled under `SyntaxShape`, which deals with both parsing (via `Expression`, `MathExpression`, etc) and semantics (via types, i.e. `Filepath`, `Float`).  I think splitting those two into opaque internal grammar and proper public type system would make understanding the Nu language easier.

**Comment by ItsHarper** [comment]
> One thing I'd like to add is that Nushell already has poor man's sum types in form of OneOf.

This is great info that needs documentation (though it's actually `oneof`)

**Comment by nome** [comment]
Apart from [theoretical considerations](https://www.haskellforall.com/2022/06/the-appeal-of-bidirectional-type.html), one practical benefit of leaning towards [TypeScripts' type system](https://www.thestrangeloop.com/2018/understanding-typescripts-structural-type-system.html) (which would probably mean bidirectional type checking instead of Hindley-Milner) would be that the basic ideas are probable more familiar to people coming to nu for its structured data concepts (which are heavily influenced by Javascript/JSON). Secondly, as @htngr pointed out, it has the benefit of proven compatibilty with gradual typing. And thirdly, I guess it should play nicely with nu's existing pattern matching and guard semantics - although I'm not a subject matter expert, so I could be wrong here.

In this spirit, here is another idea nu could steal from TypeScript: String literals as types. For example:
```nu
def greet [who: record<kind: "person", name: string>] {
  $"Hello, ($who.name)!"
}

# this is allowed:
greet {kind: "person", name: "Buzz Aldrin"}

# this causes a type checking error:
greet {kind: "teapot", name: "Russell"}
```

I'm actually experimenting with some code right now that could benefit from this (and also: type aliases! #13484).

Slightly more sophisticated example:

```nu
type Circle = record<kind: "circle", radius: number>
type Rectangle = record<kind: "rect", width: number, height: number>
type Shape = oneof<Circle, Rectangle>

def area [shape: Shape]: nothing -> number {
  match $shape {
    {kind: "circle", radius: $r} => ($math.PI * $r * $r),
    {kind: "rect", width: $w, height: $h} => ($w * $h)
  }
}
```

Note that `Circle`, `Rectangle` and `Shape` would be just aliases, type checking is still purely structural. The type checker could then, in principle, also check whether the match expression is exhaustive and flag in error if - say - `Shap` gains another variant later on.

Side note: another detail that currently seems to be missing is an ability to declare `byte stream` input/output signatures as such.

**Comment by likern** [comment]
First I would like to incline towards first-class types, which can be values themselves. So that type would be just `let Shape = type {}`. Otherwise nushell might end up to implement additional metaprogramming language just for types - as it happened for C++ with their type metaprogramming.

We definitely will want to work with types and values to perform some modifications with them (instead of constantly duplicating type definitions with slight modifications)

**Comment by nome** [comment]
> > One thing I'd like to add is that Nushell already has poor man's sum types in form of OneOf.
> 
> This is great info that needs documentation (though it's actually `oneof`)

Yes: See https://github.com/nushell/nushell.github.io/pull/2042

**Comment by likern** [comment]
I would like to take this challenge and provide rigour type system. But, since I didn't study this in university,ot will take time to study type theory. Say, one semester.

But it depends how things will be going in this project. I'm interested also because it looks like cool language for relational databases (that's my personal interest). So my personal interests intersect with nushell type system and nu language as itself.

If nu can be relaxed in it's behaviour (making it relational logical language) it means we can massively parallelize execution and apply all theory and algorithms from DBMS area.
I mean reorder operations (beyond the pipe operator) to make it more efficient without changing semantics (behaviour, results).

**Comment by danrasmuson** [comment]
I’ve fallen in love with nushell. Also at the same time, as my scripts expand to a few hundred lines I start to find the current day limitations in the type system to be a hurdle. My current needs include using the OR operators in function return types, ability to store types into variables for reuse, and ability to type functions which can intentionally throw errors.

As has been stated in the thread so far adopting Typescript standards for typing would be ideal in my eyes.

I can definitely grok how enhancing the type system is not a small feat, but it would be a great achievement for the language of Nushell. This achievement would support larger projects adopting Nushell and in turn grow its success.

**Comment by quenbyako** [comment]
Any updates about this? 🥹
